std/assembly/math.ts

```diff
@@ -2287,6 +2287,114 @@ export namespace NativeMathf {
   }
 }
 
+export namespace SafeMath {
+
+  @inline
+  export function clz<T>(x: T): T {
+    if (isInteger<T>()) {
+      return builtin_clz<T>(x);
+    }
+    throw new Error("Unexpected generic type");
+  }
+
+  @inline
+  export function clz32<T>(x: T): T {
+    if (isInteger<T>() && sizeof<T>() <= 4) {
+      return builtin_clz<T>(x);
+    }
+    throw new Error("Unexpected generic type");
+  }
+
+  @inline
+  export function min<T>(x: T, y: T): T {
+    if (isInteger<T>()) {
+      return builtin_min<T>(x, y);
+    }
+    throw new Error("Unexpected generic type");
+  }
+
+  @inline
+  export function max<T>(x: T, y: T): T {
+    if (isInteger<T>()) {
+      return builtin_max<T>(x, y);
+    }
+    throw new Error("Unexpected generic type");
+  }
+
+  @inline
+  export function abs<T>(x: T): T {
+    if (isInteger<T>()) {
+      return builtin_abs<T>(x);
+    }
+    throw new Error("Unexpected generic type");
+  }
+
+  @inline
+  export function imul<T>(x: T, y: T): T {
+    if (isInteger<T>()) {
+      return x * y;
+    }
+    throw new Error("Unexpected generic type");
+  }
+
+  @inline
+  export function pow<T>(x: T, y: i32): T {
+    if (isInteger<T>()) {
+      if (sizeof<T>() == 8) return <T>ipow64(<i64>x, y);
+      if (sizeof<T>() <= 4) return <T>ipow32(<i32>x, y);
+    }
+    throw new Error("Unexpected generic type");
+  }
+
+  @inline
+  export function sign<T>(x: T): T {
+    if (isInteger<T>()) {
+      if (!x) return <T>0;
+      if (isSigned<T>()) {
+        return 1 - (x >> (sizeof<T>() * 4 - 1) << 1);
+      } else {
+        return <T>1;
+      }
+    }
+    throw new Error("Unexpected generic type");
+  }
+
+  @inline
+  export function signbit<T>(x: T): bool {
+    if (isInteger<T>()) {
+      if (!x) return <T>0;
+      if (isSigned<T>()) {
+        return <bool>(x >> (sizeof<T>() * 4 - 1));
+      } else {
+        return <T>1;
+      }
+    }
+    throw new Error("Unexpected generic type");
+  }
+
+  @inline
+  export function sqrt<T>(x: T): T {
```

Comment:
I think this might benefit from a link to algorithm used.

Reply:
I'm not sure about integer `sqrt`. Is it necessary?
Also I could provide  integer `log2` which could be more useful I guess